src/internal/snapshot/recovery.go

```diff
@@ -214,6 +215,65 @@ func (s *Snapshotter) dumpTx(ctx context.Context, tx *pachsql.Tx, opts CreateSna
 	return id, fh, nil
 }
 
+func (s *Snapshotter) pinDatabaseDump(ctx context.Context, fh io.Reader, id SnapshotID) (retErr error) {
+	ctx, done := log.SpanContext(ctx, "pinDatabaseDump")
+	defer done(log.Errorp(&retErr))
+
+	var closedFileSet, pinnedFileSet bool
+	fw := s.Storage.NewWriter(ctx)
+	defer func() {
+		if closedFileSet {
+			return
+		}
+		if _, err := fw.Close(); err != nil {
```

Comment:
You shouldn't close the fileset if an error occurs while adding the file. Only the context needs to be canceled when an error occurs. The defer and `closedFileSet` variable can be removed. Just close after successfully adding the file.